src/infill/SpaghettiInfill.cpp

```diff
@@ -0,0 +1,139 @@
+                pillar.layer_count++;
+            }
+        }
+
+
+        // handle finished pillars
+        for (auto it = pillar_base.begin(); it != pillar_base.end();)
+        {
+            InfillPillar& pillar = *it;
+            if (pillar.layer_count >= spaghetti_max_layer_count
+                || pillar.last_layer_added < static_cast<int>(layer_idx)
+            )
+            {
+                pillar.addToTopSliceLayerPart(layer_height_mm, filling_area_inset, line_width);
+                auto to_be_erased = it;
+                ++it;
+                pillar_base.erase(to_be_erased);
+            }
+            else
+            {
+                ++it;
+            }
+        }
+    }
+    // handle unfinished pillars
+    for (auto it = pillar_base.begin(); it != pillar_base.end(); ++it)
+    {
+        it->addToTopSliceLayerPart(layer_height_mm, filling_area_inset, line_width);
+    }
+}
+
+void SpaghettiInfill::InfillPillar::addToTopSliceLayerPart(double layer_height_mm, coord_t filling_area_inset, coord_t line_width)
+{
+    SliceLayerPart& slice_layer_part = *top_slice_layer_part;
+    double volume = total_area_mm2 * layer_height_mm;
+    assert(volume > 0.0);
+
+    // get filling area
+    Polygons filling_area = top_part.offset(-filling_area_inset);
+    assert(top_part.size() > 0 && top_part[0].size() > 0 && "the top part must be a non-zero area!");
+    if (filling_area.size() == 0)
+    {
+        AABB aabb(top_part);
+        Point inside = (aabb.min + aabb.max) / 2;
+        if (!top_part.inside(inside))
+        {
+            inside = top_part[0][0];
+        }
+        filling_area = PolygonsPart();
+        PolygonRef poly = filling_area.newPoly();
+        poly.emplace_back(inside + Point(-line_width / 2 - 10, line_width / 2 + 10));
+        poly.emplace_back(inside + Point(line_width / 2 + 10, line_width / 2 + 10));
+        poly.emplace_back(inside + Point(line_width / 2 + 10, -line_width / 2 - 10));
+        poly.emplace_back(inside + Point(-line_width / 2 - 10, -line_width / 2 - 10));
+    }
+    slice_layer_part.spaghetti_infill_volumes.emplace_back(top_part, volume);
+}
+
+bool SpaghettiInfill::InfillPillar::isConnected(const PolygonsPart& infill_part) const
+{
+    Polygons insetted = infill_part.offset(-connection_inset_dist);
+    if (insetted.intersection(top_part).size() > 0)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+SpaghettiInfill::InfillPillar& SpaghettiInfill::addPartToPillarBase(const PolygonsPart& infill_part, std::list<SpaghettiInfill::InfillPillar>& pillar_base, coord_t connection_inset_dist)
+{
+    std::list<SpaghettiInfill::InfillPillar>::iterator ret = pillar_base.end();
+    for (auto it = pillar_base.begin(); it != pillar_base.end(); ++it)
+    {
+        InfillPillar& pillar = *it;
+        if (pillar.isConnected(infill_part))
+        {
+            pillar.total_area_mm2 += INT2MM(INT2MM(infill_part.area()));
```

Comment:
You later take this area and multiply it by the layer height. If taking initial layer height into account, this property should probably hold the volume rather than the area.